### ISSUE TITLE
JDBI cannot be used with JPMS with Gradle

### ISSUE THREAD

**Issue by kelemen** [jdbi/jdbi#2698]
MANIFEST.MF in the JDBI jars should be the 1st or the 2nd entry for `JarInputStream` to always work. However, it is not for JDBI. This causes Gradle to not detect JDBI as JPMS module. This issue would also be solved by an explicit `module-info.java` where there is no similar assumption.

See <https://github.com/gradle/gradle/issues/30589> for reference.

**Comment by stevenschlansker** [comment]
Hi @kelemen , I looked briefly into this and Jdbi simply uses the `maven-jar-plugin` out of the box to build its jars - nothing special. I did not see any configuration in the jar plugin around where in the jar entry order the manifest file is placed.

This might be out of the control of Jdbi entirely. Do you know how to fix this? If not, our only recourse might be to file a bug on `maven-jar-plugin` and have it fixed upstream.

**Comment by kelemen** [comment]
I looked at the pom of JDBI as well, and came to the same conclusion. So, the Maven plugin is probably wrong (I'm not aware of a workaround besides moving the build to Gradle which I'm not sure you want to make). Technically JDBI could workaround this issue by providing a `module-info.java` (I guess it is not more frequent than currently is, because not many are using automatic modules nowadays). I briefly looked at the issues about `module-info.java`, and I didn't quite understand the main reason it is hard to add (though I most likely missed something), because as far as optional dependencies go, JDBI could just declare them as `requires static` (which is the standard solution for this).

It is quite an awkward situation now, because Gradle guys immediately closed the issue (even though this is technically a bug in Gradle not in JDBI) with basically the reasoning that many tools have this bug, so they can have it too :). And sadly, my experience is that once an issue is closed, they will never look at it again.

In the following few weeks I will not really have time for anything, but after that I'm planning to submit a PR for Gradle (since it is actually trivial to fix in Gradle, and would even make it more efficient). The only thing I worry about that they might double down on wanting to keep Gradle buggy, even with a PR.

**Comment by stevenschlansker** [comment]
We are interested in providing `module-info.java` but I do not really understand / recall the blocking reasons either. We can pick up that discussion on the issues already open to add modules.

If "many tools have this bug", maybe it is worth opening an issue on Maven to fix this for all projects?

**Comment by kelemen** [comment]
I have opened an issue in Maven's JIRA, maybe they will do something about it. We will see, though actually I'm really hoping for `module-info.java` for JDBI (because without it, there are other inconveniences).

See: <https://issues.apache.org/jira/browse/MNG-8266>

**Comment by stevenschlansker** [comment]
By commenting out the `inline-maven-plugin` use, the jar order comes back to the expected ordering.

**Comment by hgschmie** [comment]
Current jar contents:

```
    0 Mon Aug 19 13:18:34 PDT 2024 META-INF/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/org.antlr/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/org.antlr/antlr4-runtime/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.antlr/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.antlr/antlr4-runtime/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.jdbi/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.jdbi/jdbi3-core/
     0 Mon Aug 19 13:18:34 PDT 2024 org/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/annotation/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/annotation/internal/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/argument/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/argument/internal/
...
```

with inline plugin 1.5.0 (soon to be released

```
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/
   780 Mon Aug 19 20:18:34 PDT 2024 META-INF/MANIFEST.MF
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/org.antlr/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/inlined/org.antlr/antlr4-runtime/
  1759 Mon Sep 04 22:10:08 PDT 2023 META-INF/inlined/org.antlr/antlr4-runtime/MANIFEST.MF
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.antlr/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.antlr/antlr4-runtime/
    59 Mon Sep 04 22:10:08 PDT 2023 META-INF/maven/org.antlr/antlr4-runtime/pom.properties
  3552 Mon Sep 04 22:10:08 PDT 2023 META-INF/maven/org.antlr/antlr4-runtime/pom.xml
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.jdbi/
     0 Mon Aug 19 13:18:34 PDT 2024 META-INF/maven/org.jdbi/jdbi3-core/
    63 Mon Aug 19 20:18:34 PDT 2024 META-INF/maven/org.jdbi/jdbi3-core/pom.properties
  5986 Mon Aug 19 20:18:34 PDT 2024 META-INF/maven/org.jdbi/jdbi3-core/pom.xml
     0 Mon Aug 19 13:18:34 PDT 2024 org/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/
     0 Mon Aug 19 13:18:34 PDT 2024 org/jdbi/v3/core/
  3677 Mon Aug 19 20:18:34 PDT 2024 org/jdbi/v3/core/AbstractHandleSupplier.class
   544 Mon Aug 19 20:18:34 PDT 2024 org/jdbi/v3/core/CloseException.class
...
```

**Comment by kelemen** [comment]
That is good news. Do you have an expected time when a new JDBI will be released with this fix?

**Comment by hgschmie** [comment]
I am pushing the build change right now. We can do a release right afterwards, which is basically 3.45.4 unchanged (there is one dependency update) and this fix.

**Comment by stevenschlansker** [comment]
Fixed in 3.46.0

**Comment by kelemen** [comment]
Thanks. Just to confirm: I have tested 3.46.0 in our build, and it works fine now.
